ADBDEV-4726-36 Check that pstate is not null

dnskvlnk opened the pull request:
Check that pstate is not null

coerce_type passes pstate where it's dereferenced without checking that it's not
null. We can't guarantee that pstate is always valid. This patch adds null-check

RekGRpth commented on src/backend/parser/parse_relation.c:
```diff
@@ -485,6 +485,8 @@ GetRTEByRangeTablePosn(ParseState *pstate,
 					   int varno,
 					   int sublevels_up)
 {
+	Assert(pstate || sublevels_up == 0);
```
```c
pstate = NULL;
sublevels_up = 0;
Assert(NULL || true);
...
return rt_fetch(varno, pstate->p_rtable); <--- segfault
```

dnskvlnk replied:
Seems like having sublevels_up == 0 doesn't have sense, thanks. Switched to Insist

RekGRpth commented on src/backend/parser/parse_relation.c:
```diff
@@ -485,6 +485,8 @@ GetRTEByRangeTablePosn(ParseState *pstate,
 					   int varno,
 					   int sublevels_up)
 {
+	Insist(pstate);
```
pstate may be NULL here https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/parser/parse_coerce.c#L404

dnskvlnk replied:
pstate is checked to be non-null before it's passed to `GetRTEByRangeTablePosn`, so I think we need this check. However, it can be replaced by assert, what do you think?

https://github.com/arenadata/gpdb/blob/0f972b1201828277979669744ab7bd007e623ea3/src/backend/parser/parse_coerce.c#L627-L630

RekGRpth replied:
> pstate is checked to be non-null before it's passed to `GetRTEByRangeTablePosn`

What does the coerce_unknown_var function have to do with it? I was talking about the coerce_type function.

dnskvlnk replied:
And it's ok unless you point to a place where it's dereferenced
https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/parser/parse_coerce.c#L155

RekGRpth replied:
Maybe we should have added a check here?
https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/parser/parse_coerce.c#L540-L546
```c
	if (inputTypeId == RECORDOID &&
		ISCOMPLEX(targetTypeId) &&
		PointerIsValid(pstate))
	{
		/* Coerce a RECORD to a specific complex type */
		return coerce_record_to_complex(pstate, node, targetTypeId,
										ccontext, cformat, location);
	}
```

dnskvlnk replied:
Changed condition to this, but used  != NULL to stick with exsisting style

RekGRpth commented on src/backend/parser/parse_relation.c:
```diff
@@ -1355,7 +1353,8 @@ addRangeTableEntryForFunction(ParseState *pstate,
 		 * used as a RangeTableEntry.
 		 */
-		if (funcexpr && IsA(funcexpr, FuncExpr))
+		Assert(funcexpr);
```
Is this patch about pstate?

dnskvlnk replied:
Wrong branch, do you mind if I make force push?

RekGRpth replied:
yes, you may force-push this branch

dnskvlnk replied:
Done

Commit:
Check that pstate is not null

coerce_type passes pstate where it's dereferenced without checking that it's not
null. We can't guarantee that pstate is always valid. This patch adds null-check
